_posts/2019-01-23-multifile-python-functions.md

```diff
@@ -0,0 +1,213 @@
+2019/01/19 10:55:34 Attempting to expand templates from https://github.com/openfaas-incubator/python-flask-template
+2019/01/19 10:55:35 Fetched 3 template(s) : [python27-flask python3-flask python3-flask-armhf] from https://github.com/openfaas-incubator/python-flask-template
+$ faas-cli new wordcloud --lang python3-flask
+$ mv wordcloud.yml stack.yml
+```
+
+The project should now look like
+
+```txt
+.
+├── stack.yml
+├── template
+│   ├── python27-flask
+│   │   ├ ...
+│   ├── python3-flask
+│   │   ├ ...
+│   └── python3-flask-armhf
+│       ├ ...
+└── wordcloud
+    ├── __init__.py
+    ├── handler.py
+    └── requirements.txt
+```
+
+## The word count implementation with relative imports
+
+We are going to add two more files to our function:
+
+```sh
+touch wordcloud/stopwords
+touch wordcloud/wordcloud.py
+```
+
+The project now looks like
+
+```txt
+.
+├── stack.yml
+├── template
+│   ├── ...
+└── wordcloud
+    ├── __init__.py
+    ├── handler.py
+    ├── requirements.txt
+    ├── stopwords
+    └── wordcloud.py
+```
+
+The `stopwords` is a plain text file of words that will be excluded in the wordcount. These are short common words that you would not want to include in a wordcloud visualization such as: `a`, `an`, `him`, `her`, and so on. This list of words will depend on your use case, adjust as you see fit.
+
+All of the real fun happens in `wordcloud.py`:
+
+```py
+# modified wordcloud.py
+import unicodedata
+import os
+from typing import Dict, List
+
+from operator import itemgetter
+from collections import defaultdict
+
+FILE = os.path.dirname(__file__)
+STOPWORDS = set(
+    map(str.lower,
+        map(str.strip,
+            open(os.path.join(FILE, 'stopwords'), 'r').readlines()
+            )
+        )
+)
+
+
+def process_text(text: bytes) -> Dict[str, int]:
+    """Splits a long text into words a count of interesting words in the text.
+
+    process_text will eliminate any of the stopwords, punctuation, and normalize
+    the text to merge cases and plurals into a single value.
+    """
+    words = text.decode("utf-8").split()
+    # remove stopwords
+    # remove 's
+    words = [
+        word[:-2] if word.lower().endswith(("'s",))
+        else word
+        for word in words
+    ]
+    # remove numbers
+    words = [word for word in words if not word.isdigit()]
+    words = [strip_punctuation(word) for word in words]
+    words = [word for word in words if word.lower() not in STOPWORDS]
+
+    return process_tokens(words)
+
+def process_tokens(words: List[str]) -> Dict[str, int]:
+    """Normalize cases and remove plurals.
+    """
+    # ...
+
+def strip_punctuation(text: str) -> str:
+    # ...
+```
+
+Note that processing the `STOPWORDS` at the start of the file means it will be loaded into memory once the package is imported. In this case, once OpenFaaS deploys the function.
+
+Using an explicit relative import, we can very easily use the `process_text` method to create a very simple `handler.py`:
```

Comment:
`explicit relative import` is jargon, can you link to this concept in the Python docs?

Reply:
The language come from PEP 8 https://www.python.org/dev/peps/pep-0008/

> Absolute imports are recommended, as they are usually more readable and tend to be better behaved (or at least give better error messages) if the import system is incorrectly configured (such as when a directory inside a package ends up on sys.path):
> 
> import mypkg.sibling
> from mypkg import sibling
> from mypkg.sibling import example
> However, explicit relative imports are an acceptable alternative to absolute imports, especially when dealing with complex package layouts where using absolute imports would be unnecessarily verbose:
> 
> from . import sibling
> from .sibling import example
> Standard library code should avoid complex package layouts and always use absolute imports.
> 
> Implicit relative imports should never be used and have been removed in Python 3.

Unfortunately, it doesn't have a directly link anchor.

This distinction will make sense to anyone that been using python for awhile. *BUT* given that they "have been removed in Python 3" I will just remove `explicit` from the blog post.